sesame2spiner/parser.cpp

```diff
@@ -12,103 +10,63 @@
 // paid-up, irrevocable worldwide license in this material to reproduce,
 // prepare derivative works, distribute copies to the public, perform
 // publicly and display publicly, and to permit others to do so.
-//======================================================================
+//------------------------------------------------------------------------------
 
-#include <regex>
-#include <string>
-#include <sstream>
-#include <cstring>
-#include <iostream>
-#include <cstdlib>
+#include <algorithm>
 #include <fstream>
+#include <istream>
+#include <sstream>
+#include <stdexcept>
+#include <string>
+#include <unordered_map>
 
-#include <nlohmann/json.hpp>
 #include "parser.hpp"
-#include "io_eospac.hpp"
 
-std::string removeComments(const std::string& in) {
-  using std::regex;
-  using std::regex_replace;
-
-  // first part is /* comment */. Second part is // comment
-  // Source:
-  // https://blog.ostermiller.org/find-comment
-  regex comments("(/\\*([^*]|[\r\n]|(\\*+([^*/]|[\r\n])))*\\*+/)|(//(.*))");
-
-  return regex_replace(in,comments,"");
+Params::Params(const std::string &input_file) {
+  std::ifstream config_file(input_file);
+  if (config_file.is_open()) {
+    Parse(config_file);
+  } else {
+    throw std::runtime_error("Couldn't open config file " + input_file + "\n");
+  }
 }
 
-nlohmann::json stringToJson(const std::string& in) {
-  return nlohmann::json::parse(removeComments(in));
+void Params::Parse(std::istream &s) {
+  std::string line;
+  while (getline(s, line)) {
+    line.erase(std::remove_if(line.begin(), line.end(), isspace), line.end());
+    if (line[0] == '#' || line.empty())
+      continue;
+    auto delimiter_pos = line.find("=");
+    auto name = line.substr(0, delimiter_pos);
+    auto value = line.substr(delimiter_pos + 1);
+    params_[name] = value;
+  }
 }
 
-nlohmann::json fileToJson(const std::string& filename) {
-  std::ifstream file(filename.c_str());
-  std::stringstream buffer;
-  if (!file.good()) {
-    std::cerr << "Cannot read file " << filename << std::endl;
-    std::exit(1);
-  }
-  buffer << file.rdbuf();
-  return stringToJson(buffer.str());
+template <> std::string Params::Get(const std::string &key) const {
+  return params_.at(key);
```

Comment:
`unordered_map::at` gives an obscure (at least to me) error if `key` isn't contained. What about adding a check of `Params::Contains` here first, and if the parameter isn't contained, some helpful message like `"Parameter \"" + key + "\" not found!"`?

Reply:
Yeah that's a good idea.